File: python/snake_game/game.py
"""game state machine + fixed-timestep loop. renders every frame, steps snake via accumulator"""
import sys

from . import config
from .assets import load_sprites
from .audio_manager import AudioManager
from .food import FoodManager
from .highscore import HighScoreStore
from .input_handler import InputHandler
from .menu import GameSettings, MenuController
from .renderer import Renderer
from .snake import RIGHT, Snake
from .ai.snake_ai import SnakeAI

START_CELL = (3, 5)


class Game:
    def __init__(self, start_with_ai=False):
        import pygame  # deferred, so import doesn't need a display
        self._pygame = pygame

        pygame.init()
        pygame.display.set_caption('Snake')
        pygame.display.set_icon(pygame.image.load('assets/snake_game_icon.jpg'))

        self.surface = pygame.display.set_mode((config.WINDOW_WIDTH, config.WINDOW_HEIGHT))
        self.font = pygame.font.SysFont('Arial', 30)
        self.clock = pygame.time.Clock()

        self.sprites = load_sprites()
        self.audio = AudioManager()
        self.high_score = HighScoreStore(config.HIGH_SCORE_FILE)
        self.input = InputHandler()
        self.renderer = Renderer(self.surface, self.sprites, self.font)
        self.ai = SnakeAI(config.GRID_COLS, config.GRID_ROWS)

        self.settings = GameSettings(ai_enabled=start_with_ai)
        self.menu = MenuController(self.surface, self.settings)

        self.last_score = 0

    def run(self):
        try:
            self.menu.show_start()
            while True:
                self._play_round()
                self.menu.show_game_over(self.last_score, self.high_score.value)
        finally:
            self._pygame.quit()

    def _play_round(self):
        pygame = self._pygame
        snake = Snake(START_CELL, RIGHT)
        food = FoodManager(config.GRID_COLS, config.GRID_ROWS, count=self.settings.apple_count)
        food.reset(snake.occupied_cells())

        ai_active = self.settings.ai_enabled
        use_hamiltonian = False
        started = ai_active  # human needs a keypress first, AI doesn't wait

        accumulator = 0.0
        step_seconds = 1.0 / config.LOGIC_TICKS_PER_SECOND

        while True:
            frame_ms = self.clock.tick(config.RENDER_FPS)
            accumulator += frame_ms / 1000.0

            for command, value in self.input.poll():
                if command == 'quit':
                    pygame.quit()
                    sys.exit(0)
                elif command == 'toggle_ai':
                    ai_active = not ai_active
                    started = True
                elif command == 'toggle_ai_mode':
                    use_hamiltonian = not use_hamiltonian
                elif command == 'direction' and not ai_active:
                    snake.set_direction(value)
                    started = True

            while accumulator >= step_seconds:
                accumulator -= step_seconds
                if not started:
                    continue

                if ai_active:
                    direction = self.ai.decide(snake, food.items, use_hamiltonian)
                    if direction:
                        snake.set_direction(direction)

                next_head = snake.peek_next_head()
                eating = next_head in food.items
                if eating:
                    snake.grow()
                snake.advance()

                if snake.hits_wall(config.GRID_COLS, config.GRID_ROWS) or snake.hits_self():
                    self.audio.play_death()
                    self.last_score = snake.length - 1
                    self.high_score.update(self.last_score)
                    return

                if eating:
                    self.audio.play_eat()
                    food.consume(next_head, snake.occupied_cells())

            self._render(snake, food, ai_active)

    def _render(self, snake, food, ai_active):
        self.renderer.draw_board()
        self.renderer.draw_food(food.items)
        self.renderer.draw_snake(snake)
        self.renderer.draw_hud(snake.length - 1, self.high_score.value, ai_active)
        self._pygame.display.flip()

File: cpp/src/game.cpp
#include "game.hpp"

#include <raylib.h>

#include <algorithm>
#include <random>
#include <unordered_set>

#include "config.hpp"

namespace {

std::unordered_set<Cell, CellHash> ToSet(const std::deque<Cell>& body) {
    return std::unordered_set<Cell, CellHash>(body.begin(), body.end());
}

}  // namespace

Game::Game(bool startWithAi, std::optional<std::string> recordPath, std::optional<std::string> replayPath)
    : window_(config::WindowWidth, config::WindowHeight, "Snake"),
      sprites_("assets/snake_spritesheet.png"),
      trophy_("assets/trophy.png"),
      highScore_(config::HighScoreFile),
      renderer_(sprites_, trophy_),
      ai_(config::GridCols, config::GridRows),
      recordPath_(std::move(recordPath)),
      replayPath_(std::move(replayPath)) {
    SetTargetFPS(config::RenderFps);

    Image icon = LoadImage("assets/snake_game_icon.jpg");
    ImageFormat(&icon, PIXELFORMAT_UNCOMPRESSED_R8G8B8A8);
    SetWindowIcon(icon);
    UnloadImage(icon);

    sprites_.Define("apple", Rectangle{0, 192, 64, 64});
    sprites_.Define("head_left", Rectangle{192, 64, 64, 64});
    sprites_.Define("head_right", Rectangle{256, 0, 64, 64});
    sprites_.Define("head_up", Rectangle{192, 0, 64, 64});
    sprites_.Define("head_down", Rectangle{256, 64, 64, 64});
    sprites_.Define("body", Rectangle{64, 0, 64, 64});

    settings_.aiEnabled = startWithAi;
}

void Game::Run() {
    if (replayPath_) {
        PlayReplay(*replayPath_);
        return;
    }

    if (!menu_.ShowStart(settings_)) return;

    while (true) {
        PlayRound();
        if (!menu_.ShowGameOver(lastScore_, highScore_.Value())) return;
    }
}

void Game::PlayRound() {
    const Cell startCell{3, 5};
    const Direction startDirection = Direction::Right;
    const unsigned int foodSeed = std::random_device{}();

    Snake snake(startCell, startDirection);
    FoodManager food(config::GridCols, config::GridRows, settings_.appleCount, foodSeed);
    food.Reset(ToSet(snake.Body()));
    std::vector<Direction> recordedDirections;

    bool aiActive = settings_.aiEnabled;
    bool useHamiltonian = false;
    bool showDebug = false;
    bool started = aiActive;  // human needs a keypress first, AI doesn't wait

    double accumulator = 0.0;
    const double stepSeconds = 1.0 / config::LogicTicksPerSecond;
    std::deque<Cell> previousBody = snake.Body();

    while (!WindowShouldClose()) {
        accumulator += GetFrameTime();

        for (const Command& command : input_.Poll()) {
            switch (command.type) {
                case CommandType::Quit:
                    return;
                case CommandType::ToggleAi:
                    aiActive = !aiActive;
                    started = true;
                    break;
                case CommandType::ToggleAiMode:
                    useHamiltonian = !useHamiltonian;
                    break;
                case CommandType::ToggleDebug:
                    showDebug = !showDebug;
                    break;
                case CommandType::Direction:
                    if (!aiActive) {
                        snake.SetDirection(command.direction);
                        started = true;
                    }
                    break;
            }
        }

        while (accumulator >= stepSeconds) {
            accumulator -= stepSeconds;
            if (!started) continue;

            // snapshot pre-tick body so Render() can interpolate to it
            previousBody = snake.Body();

            if (aiActive) {
                auto direction = ai_.Decide(snake, food.Items(), useHamiltonian);
                if (direction) snake.SetDirection(*direction);
            }

            Cell nextHead = snake.PeekNextHead();
            const auto& items = food.Items();
            bool eating = std::find(items.begin(), items.end(), nextHead) != items.end();
            if (eating) snake.Grow();
            snake.Advance();

            // record after Advance() so it's what actually happened, not what was requested
            if (recordPath_) recordedDirections.push_back(snake.CurrentDirection());

            if (snake.HitsWall(config::GridCols, config::GridRows) || snake.HitsSelf()) {
                audio_.PlayDeath();
                lastScore_ = snake.Length() - 1;
                highScore_.Update(lastScore_);
                if (recordPath_ && !recorded_) {
                    replay::Save(*recordPath_,
                                 ReplayData{startCell, startDirection, foodSeed, settings_.appleCount,
                                            recordedDirections});
                    recorded_ = true;
                }
                return;
            }

            if (eating) {
                audio_.PlayEat();
                food.Consume(nextHead, ToSet(snake.Body()));
            }
        }

        float alpha = static_cast<float>(accumulator / stepSeconds);
        Render(previousBody, snake, food, aiActive, showDebug, alpha);
    }
}

void Game::PlayReplay(const std::string& path) {
    ReplayData data = replay::Load(path);

    Snake snake(data.startCell, data.startDirection);
    FoodManager food(config::GridCols, config::GridRows, data.appleCount, data.foodSeed);
    food.Reset(ToSet(snake.Body()));

    std::size_t cursor = 0;
    double accumulator = 0.0;
    const double stepSeconds = 1.0 / config::LogicTicksPerSecond;
    std::deque<Cell> previousBody = snake.Body();

    while (!WindowShouldClose()) {
        accumulator += GetFrameTime();

        for (const Command& command : input_.Poll()) {
            if (command.type == CommandType::Quit) return;
        }

        while (accumulator >= stepSeconds) {
            accumulator -= stepSeconds;

            if (cursor >= data.directions.size()) return;  // replay finished

            previousBody = snake.Body();
            snake.SetDirection(data.directions[cursor++]);

            Cell nextHead = snake.PeekNextHead();
            const auto& items = food.Items();
            bool eating = std::find(items.begin(), items.end(), nextHead) != items.end();
            if (eating) snake.Grow();
            snake.Advance();

            if (snake.HitsWall(config::GridCols, config::GridRows) || snake.HitsSelf()) {
                audio_.PlayDeath();
                return;
            }

            if (eating) {
                audio_.PlayEat();
                food.Consume(nextHead, ToSet(snake.Body()));
            }
        }

        float alpha = static_cast<float>(accumulator / stepSeconds);
        Render(previousBody, snake, food, /*aiActive=*/false, /*showDebug=*/false, alpha);
    }
}

void Game::Render(const std::deque<Cell>& previousBody, const Snake& snake, const FoodManager& food,
                   bool aiActive, bool showDebug, float alpha) {
    BeginDrawing();
    renderer_.DrawBoard();
    if (showDebug && aiActive) {
        renderer_.DrawAIDebug(ai_.LastDebugInfo());
    }
    renderer_.DrawFood(food.Items());
    renderer_.DrawSnake(previousBody, snake, alpha);
    renderer_.DrawHud(snake.Length() - 1, highScore_.Value(), aiActive);
    EndDrawing();
}
